webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/campaigns/DataTypes.scala

```diff
@@ -0,0 +1,133 @@
+* documentation that, without modification of the Source Code, enables
+* supplementary functions or services in addition to those offered by
+* the Software.
+*
+* Rudder is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder.campaigns
+
+import com.normation.GitVersion
+import com.normation.GitVersion.Revision
+import com.normation.NamedZioLogger
+import org.joda.time.DateTime
+import zio.duration.Duration
+import zio.json.jsonDiscriminator
+import zio.json.jsonHint
+
+
+trait Campaign {
+  def info : CampaignInfo
+  def details : CampaignDetails
+}
+
+case class CampaignInfo (
+  id : CampaignId
+  , name : String
+  , description : String
+  , status : CampaignStatus
+  , schedule : CampaignSchedule
+  , duration: Duration
+)
+
+case class CampaignId (value : String, rev: Revision = GitVersion.DEFAULT_REV)
+
+@jsonDiscriminator("value")
+sealed trait CampaignStatus
+@jsonHint("enabled")
+case object Enabled  extends CampaignStatus
+@jsonHint("disabled")
+case object Disabled extends CampaignStatus
+@jsonHint("archived")
+case class Archived(date : DateTime) extends CampaignStatus
+
+@jsonDiscriminator("type")
+sealed trait CampaignSchedule
+
+sealed trait MonthlySchedulePosition
+case object First extends MonthlySchedulePosition
+case object Second extends MonthlySchedulePosition
+case object Third extends  MonthlySchedulePosition
+case object Last extends  MonthlySchedulePosition
+case object SecondLast extends MonthlySchedulePosition
+
+sealed trait DayOfWeek {
+  def value : Int
+}
+case object Monday extends DayOfWeek {
+  val value = 1
+}
+case object Tuesday extends DayOfWeek{
+  val value = 2
+}
+case object Wednesday extends DayOfWeek{
+  val value = 3
+}
+case object Thursday extends DayOfWeek{
+  val value = 4
+}
+case object Friday extends DayOfWeek{
+  val value = 5
+}
+case object Saturday extends DayOfWeek{
+  val value = 6
+}
+case object Sunday extends DayOfWeek{
+  val value = 7
+}
+
+@jsonHint("monthly")
+case class MonthlySchedule(monthlySchedulePosition: MonthlySchedulePosition, day : DayOfWeek, startHour : Int) extends CampaignSchedule
+@jsonHint("weekly")
+case class WeeklySchedule(day : DayOfWeek, startHour : Int) extends CampaignSchedule
+@jsonHint("one-shot")
+case class OneShot(start : DateTime) extends CampaignSchedule
+
+trait CampaignDetails
+
+case class CampaignEvent(id : CampaignEventId, campaignId : CampaignId, state : CampaignEventState, start : DateTime, end : DateTime )
+case class CampaignEventId(value : String)
+
+sealed trait CampaignEventState
```

Comment:
please use the conventional way:

- add a `def value: String` in trait
- add an `object CampaignEventState` to encapsulate case objects
- add a `def values = sealarate.IforgotTheName[CampaignEventState]`
- create a `def parse(s: String) = values.find(_ == s)`

If it breaks your json serialisation, you can have an `implicit class ToValue(c: CampaignEventState) { def value = c match { case Scheduled => "scheduled" ... }`

But it's important that the string representation of the values and their parsing is done here, and nowhere else.
Same: if you need to use string representation in db queries or whatever, use `Scheduled.value` in place. 
It's more searchable, and it's easier to see what needs to be updated if one day we need to change a value (and we prefer to not do it, but futur is complicated)